### ISSUE TITLE
Very Laggy

### ISSUE THREAD

**Issue by brandon123542** [BuildCraft/BuildCraft#3616]
Hi. I play Buildcraft with Optifine and it still lags!
I have a 2015(year) dell laptop with AMD graphics.
Does not lag without buildcraft for 1.11.2!

**Comment by AlexIIL** [comment]
What sort of lag is it? DO you get lower FPS than normal, or is your TPS low? (try running "/forge tps" to check the TPS value, it should be at 20, and the "##ms" value should be below 50 for every dimension).

Also, what blocks + tiles are in the worlds you are testing with?

**Comment by Goofybud16** [comment]
Specs: 

* Linux 4.9 (Debian)

* Mesa 13

* Radeon Fury

* Minecraft 1.11.2

* i5-46980k@4.2GHz

* 8GB RAM, Minecraft allocated 4GB

Mod list is 

* Forge 13.20.0.2304

* OpenComputers 1.7.0.28

* Buildcraft 7.99.3

* CodeChicken Lib 26.0.254

* Enderstorage 1.0

* Inventory Tweaks 1.62+beta.89.be6ffe4

* JEI 4.3.5.227

* NEI 2.2.0.216

* Reborn Core 2.14.0.90

* Tech Reborn 2.2.1.142

* Optifine 

FPS in Superflat world w/o Buildcraft installed (All default graphics except use VBOs on and 32 render distance): ~70-80. TPS is 20, tick time is ~6.5 on overworld.

FPS in Superflat world w/ Buildcraft: ~60-70FPS. TPS is 20, tick time is ~5.8 on overworld. When adding a quarry, 3 sterling engines, a chest, and some stone pipe, FPS would vary from 50-70 when near it, although it typically stays near 60-70.

**Comment by AEnterprise** [comment]
doesn't seem like BC is having that much impact on your ftp, some animations might indeed need some work to be less intensive but i wouldn't call this lag, you are still around 60 fps as you say yourself

**Comment by Goofybud16** [comment]
I'm not the original poster, I just wanted to chime in that I was noticing some minor performance drops on somewhat powerful hardware. Currently it appears that my CPU is maxed and bottlenecking my GPU, although I am going to blame some of that on Mesa 13.

**Comment by LuigiHutch** [comment]
Do /debug start the wait a min then /debug stop and post the file from the debug folder in a gist and link it here

**Comment by Goofybud16** [comment]
[With Quarry](https://gist.github.com/Goofybud16/c07413985258a88cab1d455035098ab6)

[Just buildcraft](https://gist.github.com/Goofybud16/d6e589171b4157fe57a5ead07bbff740)

[No buildcraft](https://gist.github.com/Goofybud16/75d38502adfb2ed65630a73ca25a7476)

**Comment by AEnterprise** [comment]
@Goofybud16 your tps would probably drop looking at the report. But in that profiler report the quarry (although it seems to be taking a bit of time) regular entites are taking up alot more of the ticktime, the difference that between with quarry and without BC for the entities section is only 3%

quite alot is used for moving entities and their ai's, mobspawner(s) also have quite a large portion. You might want to check if there are any animal farms and/or mobfarms that have run out of control

**Comment by AlexIIL** [comment]
Although its running at 20tps so that's probably not the issue. Can you try using the debug pie (Shift + F3) and see which segments take up more with buildcraft than without?

**Comment by Goofybud16** [comment]
W/o Buildcraft, gameRender takes ~90+% of the debug pie. In gameRender, level takes ~86% and gui takes ~10-15%. tick takes ~5% of the pie, and in tick level takes ~50% and textures ~30-40%.

With Buildcraft, gameRender takes ~85-90%. In gameRender, level takes ~70%. gui takes ~20%. tick takes ~10%, and in there textures takes ~65% while level takes ~30%.

This is in the same superflat world, on peaceful, with an empty inventory in creative mode. The only thing in view is grass, dirt, a few animals, and a village off in the distance. No Buildcraft items are present in the world/inventory in either case.

**Comment by AEnterprise** [comment]
what's your memory pressure? how much ram did you allocate to MC and how much of that is actually used?

if there are no BC items visible or placed in the world they can't realy be responsible for lagging (as your debug pie sugests, number are about the same)

**Comment by Goofybud16** [comment]
With BC installed, tick takes 5% more (doubles from 4-5% to 10+%), and tick:textures take 15% more of the pie. This is consistent and repeatable across multiple runs of Minecraft with and without Buildcraft. The only difference between them is adding/removing Buildcraft.

Memory is ~50-60% allocated once everything is loaded in and settled. I have 5.5/8GiB of system RAM used.

5% doesn't seem like a large difference, but at the same time that is doubling the time/CPU/whatever the pie measures that 'tick' takes without even placing any Buildcraft blocks.

**Comment by AEnterprise** [comment]
You would probably get about the same increases when adding more mods.

That 5.5/8 GB is for your whole system. I'm asking about MC specifically, how much ram does that have allocated and how much is used? You should see this info in the F3 screen

**Comment by AlexIIL** [comment]
We do have code that runs even if no blocks are placed in the world, so I'll try to reproduce locally.

**Comment by AlexIIL** [comment]
Also its highly unlikely that the GC would *only* happen in tick rather than in all the other sections as well.

**Comment by AEnterprise** [comment]
I know we have some code that runs but so far I have no issues and a quick look didn't show likely suspects, unless I missed something. It's all kinda spread out

**Comment by Goofybud16** [comment]
I have 4GB max set and once the world fully loads in, ~60% allocated (2202MB). The Mem category slowly grows up to ~50% every 10 seconds, then GC kicks in and drops it down to 20%.

I have 2 "packs" set up, one with a different set of mods (no TechReborn, added RF Tools, RF Dims, RF Controls, Iron chests) and I notice exactly the same thing on that pack. No BC on a superflat world w/ 32+ render distance gives a ~5% for tick in the pie, adding BC increases it to 10%+.

**Comment by Goofybud16** [comment]
The biggest change in "tick" looks to be that Textures goes from 40-50% of tick (~1-3% on the right column) to 65-80% (6-12% of right column). It goes from being 40% (and increasing to 55% or so max when turning/walking) to being 65% (and increasing to 80% or so max when turning/walking.)

**Comment by AlexIIL** [comment]
Yeah, it looks like ````AtlasSpriteSwappable```` is taking up a while to animate its texture...

**Comment by AlexIIL** [comment]
We have about 30 fluid textures that are animated, which seem to be taking up a little bit of time.

**Comment by AlexIIL** [comment]
We've currently been using a single texture for all of them, and recolouring it differently before uploading it, however I'll try using the normal colouring mechanism (it doesn't require so many textures)

**Comment by piks** [comment]
Can confirm, has been tying to find a solution for my friends, they have major client side lag while running 7.99.3 and this problem has continued into 7.99.4, we have tried turning all graphics down and can confirm when buildcraft is removed from server the lag is gone. As stated before this problem is client side. 

Temp Solution (seems to have fixed it)
Following the steps here have solved the problem for my friends moe specifically the mipmaps and the set affinity step, 
https://www.planetminecraft.com/blog/minecraft-lag-fix-no-more-lag-works/

**Comment by AEnterprise** [comment]
the guide does some things that might help reduce lag (less render distance means if it doesn't render as much so frees up render time for other things)

this is indeed something that is more of a problem on single core computers because all the rendering and the game logic has to be executed on the same core so one has to wait for the other instead of running simuntaniously, the guide also made them give the java process a higher priority so it helped with lag by making sure the process can jump the waiting queue for the processor over other applications that might be running in the background

**Comment by piks** [comment]
I still think that its strange that a mod such as buildcraft which in comparison to other mods on the server lags so much, because the amount of things buildcraft adds compaired to other mods installed are very minimal in terms of actual content, for buildcraft itself at this stage to cause this much of a difference there must be something in the mod itself thats causing something. As I can't reproduce it myself I can't investigate what.

**Comment by AlexIIL** [comment]
Can you test with this resource pack to see if it changes anything? Fluids won't animate with this: https://www.dropbox.com/s/3f7cgrs0q9l69s8/bc_fluid_still.zip?dl=1

**Comment by Goofybud16** [comment]
I tested that resource pack and observed the following:

"tick" dropped to ~5-7% while standing (still up a bit from Vanilla but still down from the 10%+ it was before. That small amount could also be from the various other mods I have installed).

"tick" dropped to ~5-10% (only rarely peaking at 10%) while moving around, which is much better than it was before (10% normal, occasionally it peaked at 20%).

So the texture pack definitely has a noticeable improvement, and will probably be a much bigger improvement on single-core or older PCs.

**Comment by AlexIIL** [comment]
@Goofybud16 thanks for testing, I'll try to add a config option to switch between a cheaper approach (but it might not look quite as good) and the current one.

**Comment by AEnterprise** [comment]
is the config still gona be added?

**Comment by AlexIIL** [comment]
At some point, when I get around to it...

**Comment by asiekierka** [comment]
For those of you experiencing the problem: try disabling mipmaps. For those of you experiencing the problem on AMD cards, try disabling mipmaps *and* installing the latest version of FoamFIx Anarchy.
